spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java

```diff
@@ -0,0 +1,186 @@
+        .filter(filter)
+        .planFiles();
+    List<DataFile> currentDataFiles =
+        Lists.newArrayList(CloseableIterable.transform(fileScanTasks, FileScanTask::file));
+    if (currentDataFiles.isEmpty()) {
+      return RewriteDataFilesActionResult.empty();
+    }
+
+    List<CloseableIterable<FileScanTask>> groupedTasks = groupTasksByPartition(fileScanTasks);
+
+    long splitSize = PropertyUtil.propertyAsLong(
+        table.properties(),
+        TableProperties.SPLIT_SIZE,
+        TableProperties.SPLIT_SIZE_DEFAULT);
+    int lookbak = PropertyUtil.propertyAsInt(
+        table.properties(),
+        TableProperties.SPLIT_LOOKBACK,
+        TableProperties.SPLIT_LOOKBACK_DEFAULT);
+    long openFileCost = PropertyUtil.propertyAsLong(
+        table.properties(),
+        TableProperties.SPLIT_OPEN_FILE_COST,
+        TableProperties.SPLIT_LOOKBACK_DEFAULT);
+
+    List<CloseableIterable<CombinedScanTask>> groupedCombinedTasks = groupedTasks.stream()
+        .map(tasks -> {
+          CloseableIterable<FileScanTask> splitTasks = TableScanUtil.splitFiles(tasks, splitSize);
+          return TableScanUtil.planTasks(splitTasks, splitSize, lookbak, openFileCost);
+        })
+        .collect(Collectors.toList());
+    CloseableIterable<CombinedScanTask> combinedScanTasks = CloseableIterable.concat(groupedCombinedTasks);
+
+    Broadcast<FileIO> io = sparkContext.broadcast(fileIO);
+    Broadcast<EncryptionManager> encryption = sparkContext.broadcast(encryptionManager);
+    boolean caseSensitive = Boolean.parseBoolean(spark.conf().get("spark.sql.caseSensitive", "false"));
+
+    RowDataRewriter rowDataRewriter = new RowDataRewriter(table, sparkContext, spec,
+        caseSensitive, io, encryption, targetDataFileSizeBytes);
+    List<DataFile> addedDataFiles = rowDataRewriter.rewriteDataForTasks(combinedScanTasks);
+
+    replaceDataFiles(currentDataFiles, addedDataFiles);
+
+    return new RewriteDataFilesActionResult(currentDataFiles, addedDataFiles);
+  }
+
+  private List<CloseableIterable<FileScanTask>> groupTasksByPartition(CloseableIterable<FileScanTask> tasks) {
+    Map<StructLike, List<FileScanTask>> tasksGroupedByPartition = Maps.newHashMap();
+
+    tasks.forEach(task -> {
+      List<FileScanTask> taskList = tasksGroupedByPartition.getOrDefault(task.file().partition(), Lists.newArrayList());
+      taskList.add(task);
+      tasksGroupedByPartition.put(task.file().partition(), taskList);
+    });
+
+    return tasksGroupedByPartition.values().stream()
+        .map(list -> CloseableIterable.combine(list, tasks))
+        .collect(Collectors.toList());
+  }
+
+  private void replaceDataFiles(Iterable<DataFile> deletedDataFiles, Iterable<DataFile> addedDataFiles) {
+    try {
+      RewriteFiles rewriteFiles = table.newRewrite();
+      rewriteFiles.rewriteFiles(Sets.newHashSet(deletedDataFiles), Sets.newHashSet(addedDataFiles));
+      commit(rewriteFiles);
+    } catch (Exception e) {
```

Comment:
We typically use the `boolean threw` pattern instead of using a broad catch like this. The purpose is to still try to clean up even if the problem wasn't an `Exception` (like `AssertionError`):

```
boolean threw = true;
try {
  RewriteFiles = table.newRewrite().rewriteFiles(...).commit()
  threw = false;
} finally {
  Tasks.foreach(...)
      .run(fileIO::deleteFile);
}
```